OptimizelySDK/DecisionService.cs

```diff
@@ -0,0 +1,331 @@
+                return true;
+
+            Logger.Log(LogLevel.INFO, string.Format("Not activating user test_user.", experiment.Key));
+            Logger.Log(LogLevel.INFO, string.Format("Experiment {0} is not running.", experiment.Key));
+
+            return false;
+        }
+
+        private bool ValidateUserProfile(string userId)
+        {
+            if (UserProfileService == null)
+                return false;
+
+            Dictionary<string, object> userProfileMap = UserProfileService.Lookup(userId);
+            return userProfileMap == null 
+                ? false 
+                : UserProfileUtils.IsValidUserProfileMap(userProfileMap);
+        }
+
+        private bool ValidateAudienceFromTargeting(Experiment experiment, UserAttributes filteredAttributes, string userId)
+        {
+            if (Validator.IsUserInExperiment(ProjectConfig, experiment, filteredAttributes))
+                return true;
+
+            Logger.Log(LogLevel.INFO, string.Format("Not activating user {0}.", userId));
+            Logger.Log(LogLevel.INFO, string.Format("User \"{0}\" does not meet conditions to be in experiment \"{1}\".", userId, experiment.Key));
+
+            return false;
+        }
+
+        private bool IsInWhiteListing(Experiment experiment, string userId)
+        {
+            return experiment.IsUserInForcedVariation(userId);
+        }
+
+        private bool CheckStickyBucketing()
+        {
+            return UserProfileService != null;
+        }
+
+        private Variation StoreAndFetchUserProfile(string userId, Experiment experiment)
+        {
+            if (UserProfileService == null)
+                return null;
+
+            UserProfile userProfile = null;
+
+            try
+            {
+                Dictionary<string, object> userProfileMap = UserProfileService.Lookup(userId);
+
+                if (userProfileMap == null)
+                    Logger.Log(LogLevel.INFO, "Unable to retrieve a UserProfile map from the UserProfileService.");
+                else if (!UserProfileUtils.IsValidUserProfileMap(userProfileMap))
+                    Logger.Log(LogLevel.ERROR, "The UserProfileService returned an invalid map.");
+                else
+                    userProfile = UserProfileUtils.ConvertMapToUserProfile(userProfileMap);
+            }
+            catch (Exception exception)
+            {
+                Logger.Log(LogLevel.ERROR, exception.Message);
+                ErrorHandler.HandleError(new Exceptions.OptimizelyRuntimeException(exception.Message));
+            }
+
+            return GetStoredVariation(experiment, userProfile ?? new UserProfile(userId, new Dictionary<string, Decision>()));
+        }
+
+        private Variation GetAndSaveVariationFromAudienceTargeting(Experiment experiment, UserAttributes filteredAttributes, string userId)
+        {
+            UserProfile userProfile = new UserProfile(userId, new Dictionary<string, Decision>());
```

Comment:
This is not correct. We don't always create a new user profile. If there is one stored in the user profile service already, then we should override its decision and re-save it.